### PR TITLE
Fix #12625: Improve poor performance of paginated treeTable with larg…

### LINKED ISSUE #12625: TreeTable: poor performance in paginated treeTable with large number of nodes
### Describe the bug

The **treeTable** component experiences significant performance degradation with large datasets when pagination is enabled. This is caused by an unbounded iteration of root nodes in the `processNode` method, which processes all nodes instead of restricting to nodes on the current page.

### Reproducer

_No response_

### Expected behavior

_No response_

### PrimeFaces edition

None

### PrimeFaces version

13.0

### Theme

_No response_

### JSF implementation

MyFaces

### JSF version

4.0

### Java version

21

### Browser(s)

_No response_

### PR DESCRIPTION
**Pull request by skybber**
I've tested it on v13.0.9

Fix #12625

### CODE REVIEW

**Review by melloware — APPROVED** [review]

**Comment by melloware** [comment]
This looks good to me but want @Rapster to review also.

**Comment by skybber** [comment]
Perhaps it would be better to have that iteration in UITree, where isPaginator() would always return false. This way, the patch code is dependent on the implementation in the superclass.

**Comment by melloware** [comment]
Well I dont think UITree can ever be paginated its only for TreeTable so i think its OK where it is?

**Comment by skybber** [comment]
If UITree is never going to be paginated, it would be better to leave it as it is and keep the patch in its current form, as you are writing.

File: primefaces/src/main/java/org/primefaces/component/treetable/TreeTableBase.java [hunk 1]
```diff
@@ -526,4 +530,32 @@ public String getExportTag() {
     public void setExportTag(String exportTag) {
         getStateHelper().put(PropertyKeys.exportTag, exportTag);
     }
-}
+
+    protected void processNode(FacesContext context, PhaseId phaseId, TreeNode root, TreeNode treeNode, String rowKey) {
```
Comment (melloware) [inline primefaces/src/main/java/org/primefaces/component/treetable/TreeTableBase.java]:
```suggestion
    @Override
    protected void processNode(FacesContext context, PhaseId phaseId, TreeNode root, TreeNode treeNode, String rowKey) {
```

**Review by melloware — CHANGES_REQUESTED** [review]

**Review by melloware — APPROVED** [review]

**Comment by Rapster** [comment]
Should be in TreeTable

**Comment by skybber** [comment]
@melloware @Rapster , moved to TreeTable

File: primefaces/src/main/java/org/primefaces/component/treetable/TreeTableBase.java [hunk 2]
```diff
@@ -25,6 +25,10 @@
 
 import javax.el.MethodExpression;
 import javax.faces.component.behavior.ClientBehaviorHolder;
+import javax.faces.context.FacesContext;
+import javax.faces.event.PhaseId;
+
+import java.util.List;
```
Comment (Rapster) [inline primefaces/src/main/java/org/primefaces/component/treetable/TreeTableBase.java]: To revert

**Comment by skybber** [comment]
@Rapster done

**Review by melloware — APPROVED** [review]

File: primefaces/src/main/java/org/primefaces/component/treetable/TreeTable.java [hunk 3]
```diff
@@ -588,4 +588,33 @@ protected boolean isCacheableColumns(List<UIColumn> columns) {
...
+                List<TreeNode> children = root.getChildren();
+                int childCount = root.getChildCount();
+                int last = (first + rows);
+                if (last > childCount) {
+                    last = childCount;
+                }
+
+                for (int i = first; i < last; i++) {
+                    TreeNode child = children.get(i);
+                    String childRowKey = rowKey == null ? String.valueOf(i) : rowKey + SEPARATOR + i;
```
Comment (Rapster) [inline primefaces/src/main/java/org/primefaces/component/treetable/TreeTable.java]: Can we introduce an util method to create the childRowKey, I saw it several times in UITree already

Comment (melloware) [reply primefaces/src/main/java/org/primefaces/component/treetable/TreeTable.java]: I assume you want this util method in the UITree itself?

Comment (Rapster) [reply primefaces/src/main/java/org/primefaces/component/treetable/TreeTable.java]: Yes

**Review by melloware — APPROVED** [review]

**Review by Rapster — APPROVED** [review]

**Pull request merged**